### PR TITLE
Adding retry logic to the `TritonInferenceStage` to allow recovering from errors

### LINKED ISSUE #1510: [FEA]: Update TritonInferenceService stage so that it is resilient and reliable to support usage in production code.
### Is this a new feature, an improvement, or a change to existing functionality?

Improvement

### How would you describe the priority of this feature request

High

### Please provide a clear description of problem this feature solves

I want Morpheus stages, specifically the TritonInferenceStage to be updated so that it meets a number of criteria required for a production service:

**Reliability**
- **Fault Tolerance and Graceful Degradation**: The system should exhibit fault tolerance by handling failures gracefully when the Triton service it depends on becomes unavailable or erratic. This includes implementing strategies for retrying requests, using circuit breakers to avoid overwhelming a failing service, and providing fallback mechanisms where appropriate to maintain functionality.
- **Error Handling and Validation**: Enhance error handling to manage and respond to various failure modes, including network issues, service downtime, and unexpected data formats or values. Implement comprehensive input and response validation to mitigate issues arising from irregular or malformed data.

**Resilience**

- **Message Durability and Recovery**: Ensure the system maintains the integrity of in-flight messages and has mechanisms in place for message recovery and reprocessing in the event of a failure. This could involve persistent queues, transactional processing, or checkpointing strategies to safeguard against message loss.
- **Service Redundancy and Load Balancing**: Enhance system resilience by implementing service redundancy, where feasible, and load balancing to distribute requests across multiple instances of the Triton service. This helps in mitigating the impact of a single instance's failure and improves overall system availability.

**API Changes**
- The current stage does not reliably support ControlMessages and needs an optimized interface that handles them exclusively.

### Describe your ideal solution

See description.

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I have searched the [open feature requests](https://github.com/nv-morpheus/Morpheus/issues?q=is%3Aopen+is%3Aissue+label%3A%22feature+request%22%2Cimprovement%2Cenhancement) and have found no duplicates for this feature request

### PR DESCRIPTION
**Pull request by cwharris**
Factor out the Triton inference logic in the Triton inference stage to a client class which can be swapped out for a fake/test client. Keep retry logic in the Stage and test using a fake/test client.

Closes #1510 

## By Submitting this PR I confirm:
- I am familiar with the [Contributing Guidelines](https://github.com/nv-morpheus/Morpheus/blob/main/docs/source/developer_guide/contributing.md).
- When the PR is ready for review, new or existing tests cover these changes.
- When the PR is ready for review, the documentation is up to date with these changes.

### CODE REVIEW

File: morpheus/_lib/include/morpheus/stages/inference_client_stage.hpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+class MORPHEUS_EXPORT IInferenceClientSession
+{
+  public:
+    virtual std::vector<TensorModelMapping> get_input_mappings(std::vector<TensorModelMapping> input_map_overrides) = 0;
+
+    virtual std::vector<TensorModelMapping> get_output_mappings(
+        std::vector<TensorModelMapping> output_map_overrides) = 0;
+
+    virtual mrc::coroutines::Task<TensorMap> infer(TensorMap&& inputs) = 0;
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/stages/inference_client_stage.hpp:55]: Comments needed on public methods/members

File: morpheus/_lib/include/morpheus/stages/inference_client_stage.hpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+};
+
+class MORPHEUS_EXPORT IInferenceClient
+{
+  public:
+    virtual std::shared_ptr<IInferenceClientSession> get_session() = 0;
+    virtual void reset_session()                                   = 0;
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/stages/inference_client_stage.hpp:62]: Comments needed on public methods/members

File: morpheus/_lib/include/morpheus/stages/inference_client_stage.hpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+     */
+    InferenceClientStage(std::unique_ptr<IInferenceClient>&& client,
+                         std::string model_name,
+                         bool needs_logits,
+                         std::vector<TensorModelMapping> input_mapping,
+                         std::vector<TensorModelMapping> output_mapping);
+
+    /**
+     * TODO(Documentation)
+     */
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/stages/inference_client_stage.hpp:100]: Fill out the documentation

File: morpheus/_lib/include/morpheus/stages/triton_inference.hpp [hunk 4]
```diff
@@ -17,141 +17,112 @@
...
- * @addtogroup stages
- * @{
- * @file
- */
+struct MORPHEUS_EXPORT TritonInferInput
+{
+    std::string name;
+    std::vector<int64_t> shape;
+    std::string type;
+    std::vector<uint8_t> data;
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/stages/triton_inference.hpp:43]: Comments needed on public methods/members

File: morpheus/_lib/include/morpheus/stages/triton_inference.hpp [hunk 5]
```diff
@@ -17,141 +17,112 @@
...
+    virtual triton::client::Error is_model_ready(bool* ready, std::string& model_name) = 0;
 
+    virtual triton::client::Error model_metadata(std::string* model_metadata, std::string& model_name) = 0;
+
+    virtual triton::client::Error model_config(std::string* model_config, std::string& model_name) = 0;
+
+    virtual triton::client::Error async_infer(triton::client::InferenceServerHttpClient::OnCompleteFn callback,
+                                              const triton::client::InferOptions& options,
+                                              const std::vector<TritonInferInput>& inputs,
+                                              const std::vector<TritonInferRequestedOutput>& outputs) = 0;
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/stages/triton_inference.hpp:67]: Comments needed on public methods/members

File: morpheus/_lib/src/stages/inference_client_stage.cpp [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+        if (m_delay > m_delay_max)
+        {
+            m_delay = m_delay_max;
+        }
+
+        co_await m_on->yield_for(m_delay);
+
+        m_delay *= 2;
+    }
+};
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/stages/inference_client_stage.cpp:149]: This would be useful as a class in a separate file so other stages could use it.

File: morpheus/_lib/src/stages/inference_client_stage.cpp [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+        {
+            // Using the `count` which is the number of rows in the inference tensors. We will check later if this
+            // doesn't match the number of rows in the dataframe (`mess_count`). This happens when the size of the
+            // input is too large and needs to be broken up in chunks in the pre-process stage. When this is the
+            // case we will reduce the rows in the response outputs such that we have a single response for each
+            // row int he dataframe.
+            // TensorMap output_tensors;
+            // buffer_map_t output_buffers;
+
+            auto lock = std::shared_lock(m_session_mutex);
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/stages/inference_client_stage.cpp:172]: Add a comment here why we are using a shared_lock and not a unique_lock.

File: morpheus/_lib/src/stages/inference_client_stage.cpp [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+
+            for (auto mapping : session->get_input_mappings(m_input_mapping))
+            {
+                if (x->memory->has_tensor(mapping.tensor_field_name))
+                {
+                    model_input_tensors[mapping.model_field_name].swap(x->get_input(mapping.tensor_field_name));
+                }
+            }
+
+            // TODO(cwharris): Break inference in to batches and attempt retries on per-batch basis.
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/stages/inference_client_stage.cpp:186]: Is this TODO still needed?

File: morpheus/_lib/stages/module.cpp [hunk 9]
```diff
@@ -15,6 +15,13 @@
  * limitations under the License.
  */
 
+#include <mrc/segment/builder.hpp>  // for Builder
+#include <mrc/segment/object.hpp>
+#include <mrc/utils/string_utils.hpp>
+#include <pybind11/attr.h>      // for multiple_inheritance
+#include <pybind11/pybind11.h>  // for arg, init, class_, module_, str_attr_accessor, PYBIND11_MODULE, pybind11
+#include <pybind11/pytypes.h>   // for dict, sequence
+// for pathlib.Path -> std::filesystem::path conversions
```
Comment (mdemoret-nv) [inline morpheus/_lib/stages/module.cpp:24]: Delete this line so the formatter can rearrange the includes properly. It wont break any comment boundaries

File: morpheus/_lib/tests/CMakeLists.txt [hunk 10]
```diff
@@ -160,4 +160,10 @@ add_morpheus_test(
     test_type_util.cpp
 )
 
+add_morpheus_test(
+  NAME triton_inference_stage
+  FILES
+    stages/test_triton_inference_stage.cpp
+)
+
```
Comment (mdemoret-nv) [inline morpheus/_lib/tests/CMakeLists.txt:168]: Put the tests in alphabetical order

File: morpheus/_lib/tests/stages/test_triton_inference_stage.cpp [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+    auto memory  = std::make_shared<morpheus::TensorMemory>(count, std::move(tensors));
+    auto table   = create_test_table_with_metadata(count);
+    auto meta    = morpheus::MessageMeta::create_from_cpp(std::move(table), 1);
+    auto message = std::make_shared<morpheus::MultiInferenceMessage>(meta, 0, count, memory);
+
+    auto triton_client           = std::make_unique<FakeTritonClient>();
+    auto triton_inference_client = std::make_unique<morpheus::TritonInferenceClient>(std::move(triton_client), "");
+    auto stage = morpheus::InferenceClientStage(std::move(triton_inference_client), "", false, {}, {});
+
+    auto on           = std::make_shared<mrc::coroutines::TestScheduler>();
```
Comment (mdemoret-nv) [inline morpheus/_lib/tests/stages/test_triton_inference_stage.cpp:309]: This is one giant block of code with no comments. Break it up into logical sections and add comments.

File: morpheus/stages/inference/triton_inference_stage.py [hunk 12]
```diff
@@ -702,28 +707,51 @@ def __init__(self,
...
         if inout_mapping is not None:
             inout_mapping_.update(inout_mapping)
+            input_mapping_.update(inout_mapping)
+            output_mapping_.update(inout_mapping)
+
+        if input_mapping is not None:
+            input_mapping_.update(input_mapping)
+
+        if output_mapping is not None:
+            output_mapping_.update(output_mapping)
```
Comment (mdemoret-nv) [inline morpheus/stages/inference/triton_inference_stage.py:736]: We need to add a branch here to detect if users specify both `inout_mapping` and either `input_mapping` or `output_mapping`. Using both should be an error and likely will be overwriting options.

File: morpheus/stages/inference/triton_inference_stage.py [hunk 13]
```diff
@@ -702,28 +707,51 @@ def __init__(self,
...
+        input_mapping_ = self._INFERENCE_WORKER_DEFAULT_INOUT_MAPPING.get(c.mode, {}).get("inputs", {})
+        output_mapping_ = self._INFERENCE_WORKER_DEFAULT_INOUT_MAPPING.get(c.mode, {}).get("outputs", {})
+
+        inout_mapping_ = {}
+        inout_mapping_.update(input_mapping_)
+        inout_mapping_.update(output_mapping_)
+
         if inout_mapping is not None:
             inout_mapping_.update(inout_mapping)
+            input_mapping_.update(inout_mapping)
```
Comment (mdemoret-nv) [inline morpheus/stages/inference/triton_inference_stage.py:729]: We should deprecate `inout_mapping` if we have added separate options for input vs output.

File: morpheus/stages/inference/triton_inference_stage.py [hunk 14]
```diff
@@ -702,28 +707,51 @@ def __init__(self,
...
             "needs_logits": needs_logits
         }
 
+        self._server_url = server_url
+        self._model_name = model_name
+        self._force_conver_inputs = force_convert_inputs
+        self._use_shared_memory = use_shared_memory
+        self._input_mapping = input_mapping_
+        self._output_mapping = output_mapping_
+        self._needs_logits = needs_logits
```
Comment (mdemoret-nv) [inline morpheus/stages/inference/triton_inference_stage.py:753]: This is better than `self._kwargs`. But we dont need both. Remove `self._kwargs`

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
This is nearly there. Needs a few improvements but most are superficial:

1. All public functions and members need to be documented using doxygen
2. I have pointed out a few places where commenting could be improved, but this PR lacks comments.
3. More tests for odd use cases need to be added to `test_triton_inference_stage.cpp`

File: morpheus/_lib/src/stages/inference_client_stage.cpp [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+
+            co_yield std::move(response);
+
+            co_return;
+
+        } catch (...)
+        {
+            auto lock = std::unique_lock(m_session_mutex);
+
+            this->m_client->reset_session();
```
[the remark this answers is not in the document]
Comment (cwharris) [reply morpheus/_lib/src/stages/inference_client_stage.cpp:231]: Something needs to keep track of what the current session is, and make that information shareable across multiple invocations `on_data`. Having the client keep track of that information is what's happening now. We could instead store the session on the stage, which might make more sense because we're also handling the locking here.

File: morpheus/_lib/src/stages/inference_client_stage.cpp [hunk 16]
```diff
@@ -169,22 +171,31 @@ mrc::coroutines::AsyncGenerator<std::shared_ptr<MultiResponseMessage>> Inference
...
 
-            auto lock = std::shared_lock(m_session_mutex);
+            if (m_session == nullptr)
+            {
+                auto lock = std::unique_lock(m_session_mutex);
+                if (m_session == nullptr)
+                {
+                    m_session = m_client->create_session();
+                }
+            }
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/stages/inference_client_stage.cpp:181]: There is a potential race condition here where multiple threads are racing to create the session. Look into moving this into the catch handler.

**Review by mdemoret-nv — APPROVED** [review]

**Review by mdemoret-nv — APPROVED** [review]
